rfcs/2022-10-12-14742-flexible-metric-tags.md

```diff
@@ -0,0 +1,152 @@
+`indexmap` set to store the tag values. This newtype will hide the implementation details of the
+underlying storage from the callers. It will also add separate methods for inserting a new tag and
+replacing a tag set with a single value. The callers of the existing `insert` function will need to
+be audited to determine which use is intended at each call site.
+
+```rust
+type TagValue = Option<String>;
+
+struct MetricTags(BTreeMap<String, indexmap::IndexSet<TagValue>>);
+
+impl MetricTags {
+    // Insert returns the value unchanged if the exact (tag,value) pair already exists,
+    // otherwise it inserts a new value for the named tag.
+    fn insert(&mut self, name: String, value: TagValue) -> Option<TagValue>;
+
+    // Replace returns all the existing values when overwriting a tag.
+    fn replace(&mut self, name: String, value: Option<TagValue>) -> Option<IndexSet<TagValue>>;
+}
+```
+
+## Rationale
+
+Changing the `MetricTags` type from an alias to a newtype wrapper allows us to provide better
+compatibility for existing uses while controlling the methods for uses that need to access all the
+values.
+
+The use of an `IndexSet` for the tag value provides us with two useful invariants:
+1. Only unique values for each tag will be stored, which prevents repeated values from showing up in
+   the output.
+1. The values can be retrieved in the order they first appeared, which allows us to trivially
+   retrieve either the first or last stored value.
+
+## Drawbacks
+
+Metrics sinks that only support a single value per tag will need to be reworked accordingly for the
+new value type. Additionally, these sinks may change their behavior for sources that have been
+producing multiple tag values.
+
+## Prior Art
+
+The Datadog agent stores metric tags as a simple set of strings, equivalent to `HashSet<String>`. It
+does some clever hashing and deduplication internally to make this work efficiently. However, it
+doesn't do anything more interesting with the tags than adding and removing whole strings, which
+does not cover all our use cases.
+
+## Alternatives
+
+The Datadog agent represents the tags as a simple set of strings, ie `HashSet<String>` or
+`BTreeSet<String>`, where the key/value implications are just an interpretation detail for
+consumers. This is by far the simplest possible storage. However, Vector needs to be able to access
+tags by name for manipulation, making this representation more challenging. Through clever use of
+`BTreeSet::range` and a wrapper type for the tags, this could be made to work, but it is unclear if
+the benefits would be worth the additional complexity.
+
+When retaining `BTreeMap` as the top-level container, there are a number of options for the value
+that would support this feature but with different semantics:
+
+1. `Vec<String>` — Retains the ordering of tags as they appear, but allows for duplicate values and
+   cannot support both bare tags and multiple values simultaneously.
```

Comment:
I'm not sure if we need to differentiate these, but an `""` _might_ could represent a bare tag.

Reply:
Perhaps. Technically, there are three cases:
1. Bare tag, `tag`
2. Tag with empty value, `tag:`
3. Tag with normal value, `tag:value`